### PR TITLE
fixing a bug in if-then-else, adding test cases

### LINKED ISSUE #340: error in if-then-else
- [x] todo: also test this with an example that has multiple statements in each branch, ending in a return.

```javascript
main contract CombinedExample {
        transaction val() returns int {
	    if ( true && ! false ) {
                return 4 + 20;
	    } else {
	        return 13;
            }
        }
        transaction main() returns int{
            int x;
            x = 9 + 0;
            x = val();
            return x;
        }
}
```
should run and produce `24` but it produces `0`. The reason is that I'm translating if-then-else incorrectly. The Yul generated for `val` is
```javascript
            function val() -> _ret_1
            {
                let _tmp_1
                let _tmp_2
                let _tmp_3
                let _tmp_4
                _tmp_3 := true
                let _tmp_5
                let _tmp_6
                _tmp_6 := false
                switch _tmp_6
                case true {
                    let _tmp_7
                    _tmp_7 := false
                    _tmp_5 := _tmp_7
                }
                case false {
                    let _tmp_8
                    _tmp_8 := true
                    _tmp_5 := _tmp_8
                }
                _tmp_4 := _tmp_5
                _tmp_2 := and(_tmp_3, _tmp_4)
                switch _tmp_2
                case true {
                    let _tmp_9
                    let _tmp_10
                    let _tmp_11
                    let _tmp_12
                    _tmp_11 := 4
                    _tmp_12 := 20
                    _tmp_10 := add(_tmp_11, _tmp_12)
                    _tmp_9 := _tmp_10
                    leave
                    _tmp_1 := _tmp_9
                }
                case false {
                    let _tmp_13
                    let _tmp_14
                    _tmp_14 := 13
                    _tmp_13 := _tmp_14
                    leave
                    _tmp_1 := _tmp_13
                }
                _ret_1 := _tmp_1
            }
```
but should be something like
```javascript
            function val() -> _ret_1
            {
                let _tmp_1
                let _tmp_2
                let _tmp_3
                let _tmp_4
                _tmp_3 := true
                let _tmp_5
                let _tmp_6
                _tmp_6 := false
                switch _tmp_6
                case true {
                    let _tmp_7
                    _tmp_7 := false
                    _tmp_5 := _tmp_7
                }
                case false {
                    let _tmp_8
                    _tmp_8 := true
                    _tmp_5 := _tmp_8
                }
                _tmp_4 := _tmp_5
                _tmp_2 := and(_tmp_3, _tmp_4)
                switch _tmp_2
                case true {
                    let _tmp_9
                    let _tmp_10
                    let _tmp_11
                    let _tmp_12
                    _tmp_11 := 4
                    _tmp_12 := 20
                    _tmp_10 := add(_tmp_11, _tmp_12)
                    _tmp_9 := _tmp_10
                    _tmp_1 := _tmp_9
                    _ret_1 := _tmp_1
                    leave
                }
                case false {
                    let _tmp_13
                    let _tmp_14
                    _tmp_14 := 13
                    _tmp_13 := _tmp_14
                    _tmp_1 := _tmp_13
                    _ret_1 := _tmp_1
                    leave
                }
            }
```
This is an error in how return variables are propagated into the branches of if-then-else blocks.

### PR DESCRIPTION
**Pull request by ivoysey**
this bug came up in the process of writing a slightly large example that used `return` from both branches of an if for a demo. this PR should fix #340.

### CODE REVIEW

File: resources/tests/GanacheTests/CombinedExample.obs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+main contract CombinedExample {
```
Comment (mcoblenz) [inline resources/tests/GanacheTests/CombinedExample.obs:1]: Please make the indentation consistent.

Comment (mcoblenz) [reply resources/tests/GanacheTests/CombinedExample.obs:1]: Also, the filename says this is an example, but it's not clear what it's an example of. I think the real answer is "everything to date," but that's not a good description in the long term. Can you choose a more explanatory name?

File: resources/tests/GanacheTests/MultiLineIfRet.obs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+main contract MultiLineIfRet {
+        transaction val() returns int {
+    	    int x = 0;
+	        if ( true && ! false ) {
+	            x = 4+20;
+		        x = 11;
```
Comment (mcoblenz) [inline resources/tests/GanacheTests/MultiLineIfRet.obs:6]: Fix indentation.

File: resources/tests/GanacheTests/MultiLineIfRetSm.obs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+main contract MultiLineIfRetSm {
+        transaction val() returns int {
```
Comment (mcoblenz) [inline resources/tests/GanacheTests/MultiLineIfRetSm.obs:2]: Fix indentation.

Comment (mcoblenz) [reply resources/tests/GanacheTests/MultiLineIfRetSm.obs:2]: Also explain what this file tests.

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 4]
```diff
@@ -247,43 +247,19 @@ object CodeGenYul extends CodeGenerator {
...
+                val neg_yul: Seq[YulStatement] = neg.flatMap(s => translateStatement(s, retVar, contractName, checkedTable))
 
-                // assign back from which ever last statement gets run, or not depending on the inductive requirements
-                val assign_back = retVar match {
-                    case Some(value) => Seq(assign1(Identifier(value), id_last))
-                    case None => Seq()
-                }
-
-                // put the pieces together into a switch statement, preceeded by the evaluation of the
+                // put the pieces together into a switch statement, proceeded by the evaluation of the
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:254]: "proceeded by" isn't right.

**Review by mcoblenz — APPROVED** [review]

File: resources/tests/GanacheTests/CombinedExample.obs (hunk 1 above)
Comment (ivoysey) [reply resources/tests/GanacheTests/CombinedExample.obs:1]: I've renamed it (and the version with return) as "CombinedBasics" -- they are tests that combine all the basics of the language, sort of everything before the first interesting feature (which is structure types).

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala (hunk 4 above)
Comment (ivoysey) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:254]: whoops! spellcheck in comments is a blessing and a curse.

**Pull request merged**